python/datasources/uhc.py

```diff
@@ -0,0 +1,122 @@
+    "Diabetes": std_col.DIABETES_PCT,
+}
+
+
+class UHCData(DataSource):
+
+    @staticmethod
+    def get_id():
+        return 'UHC_DATA'
+
+    @staticmethod
+    def get_table_name():
+        return 'uhc_data'
+
+    def upload_to_gcs(self, _, **attrs):
+        raise NotImplementedError(
+            'upload_to_gcs should not be called for UHCData')
+
+    def write_to_bq(self, dataset, gcs_bucket, **attrs):
+        df = gcs_to_bq_util.load_csv_as_dataframe_from_web(
+                BASE_UHC_URL, dtype={std_col.STATE_FIPS_COL: str}
+            )
+
+        for b in [std_col.RACE_OR_HISPANIC_COL, std_col.AGE_COL, std_col.SEX_COL]:
+            breakdown_df = self.generate_breakdown(b, df)
+
+            column_types = {c: 'STRING' for c in breakdown_df.columns}
+            for col in [std_col.COPD_PCT, std_col.DIABETES_PCT]:
+                column_types[col] = 'FLOAT'
+
+            if std_col.RACE_INCLUDES_HISPANIC_COL in breakdown_df.columns:
+                column_types[std_col.RACE_INCLUDES_HISPANIC_COL] = 'BOOL'
+
+            breakdown_df.to_csv("test_%s.csv" % b)
+            gcs_to_bq_util.add_dataframe_to_bq(
+                breakdown_df, dataset, b, column_types=column_types)
+
+    def generate_breakdown(self, breakdown, df):
+        breakdown_map = {
+            "race_and_ethnicity": UHC_RACE_GROUPS,
+            "age": UHC_AGE_GROUPS,
+            "sex": UHC_SEX_GROUPS,
+        }
+
+        output = pd.DataFrame()
+        states = df['State Name'].drop_duplicates().to_list()
+
+        for state in states:
+            for breakdown_value in breakdown_map[breakdown]:
+                output_row = {}
+                output_row['state_name'] = state
+
+                if breakdown == std_col.RACE_OR_HISPANIC_COL:
+                    output_row[std_col.RACE_CATEGORY_ID_COL] = UHC_RACE_GROUPS_TO_STANDARD[breakdown_value]
+                else:
+                    output_row[breakdown] = breakdown_value
+
+                output_row['state_fips'] = constants.STATE_NAMES_TO_FIPS[state]
+
+                for determinent in UHC_DETERMINANTS_OF_HEALTH:
+                    if breakdown_value == 'All':
+                        output_row[UHC_DETERMINANTS_OF_HEALTH[determinent]] = df.loc[
+                            (df['State Name'] == state) &
+                            (df['Measure Name'] == determinent)]['Value'].values[0]
+
+                    else:
+                        row = df.loc[
+                            (df['State Name'] == state) &
+                            (df['Measure Name'].str.contains(determinent)) &
+                            (df['Measure Name'].str.contains(breakdown_value))]
+
+                        if len(row) > 0:
+                            pct = row['Value'].values[0]
+                            if pct:
+                                output_row[UHC_DETERMINANTS_OF_HEALTH[determinent]] = pct
+
+                output = output.append(output_row, ignore_index=True)
```

Comment:
In general, it's discouraged to create an empty df and keep appending rows to it, as it is quite inefficient (eg https://stackoverflow.com/questions/13784192/creating-an-empty-pandas-dataframe-then-filling-it/56746204#56746204). An easy alternative is to just build a list of lists, where each list is a row in the df, and then construct it at once at the end while also specifying the columns, e.g.:

df = pd.DataFrame(list_of_lists, columns=columns)

Reply:
Makes sense, its definitely really slow